In [3]:
# Install required libraries
!pip install -q langchain langchain-community pypdf sentence-transformers chromadb

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.0/52.0 kB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 48.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 346.6/346.6 kB 29.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.3/23.3 MB 81.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 26.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.6/4.6 MB 124.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 68.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 554.9/554.9 kB 48.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.2/18.2 MB 80.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 71.8/71.8 kB 8.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 170.9/170.9 kB 17.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.3/61.3 kB 6.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 203.7/

In [4]:
# Upload PDF
from google.colab import files

uploaded = files.upload()

Saving NBayesLogReg.pdf to NBayesLogReg.pdf


In [5]:
# Load PDF
from langchain_community.document_loaders import PyPDFLoader

pdf_file = list(uploaded.keys())[0]

loader = PyPDFLoader(pdf_file)
documents = loader.load()

print("Pages:", len(documents))

/tmp/ipykernel_1543/1717598674.py:2: DeprecationWarning: `langchain-community` is being sunset and is no longer actively maintained. See https://github.com/langchain-ai/langchain-community/issues/674 for details and migration guidance toward standalone integration packages.
  from langchain_community.document_loaders import PyPDFLoader


Pages: 18


In [6]:
# Combine all pages into one text
text = ""

for doc in documents:
    text += doc.page_content + "\n"

print("Characters:", len(text))

Characters: 36686


In [7]:
# Chunking
from langchain_text_splitters import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=500,
    chunk_overlap=50
)

chunks = text_splitter.split_text(text)

print("Number of chunks:", len(chunks))

Number of chunks: 81


In [8]:
# Preview first 5 chunks
for i, chunk in enumerate(chunks[:5]):
    print(f"\n----- Chunk {i+1} -----\n")
    print(chunk)


----- Chunk 1 -----

CHAPTER 3
GENERATIVE AND DISCRIMINATIVE
CLASSIFIERS:
NAIVE BAYES AND LOGISTIC REGRESSION
Machine Learning
Copyright c⃝2017. Tom M. Mitchell. All rights reserved.
*DRAFT OF October 1, 2020*
*PLEASE DO NOT DISTRIBUTE WITHOUT AUTHOR’S PERMISSION*
This is a rough draft chapter intended for inclusion in the upcoming sec-
ond edition of the textbook Machine Learning, T.M. Mitchell, McGraw Hill.
You are welcome to use this for educational purposes, but do not dupli-

----- Chunk 2 -----

cate or repost it on the internet. To distribute this to others, simply point
them to http://www.cs.cmu.edu/∼tom/mlbook/NBayesLogReg.pdf. For online
copies of this and other materials related to this book, visit the web site
www.cs.cmu.edu/∼tom/mlbook.html.
Please send suggestions for improvements, or suggested exercises, to
Tom.Mitchell@cmu.edu.
1 Learning Classifiers based on Bayes Rule
Here we consider the relationship between supervised learning, or function ap-

----- Chunk 3 -----



In [9]:
from sentence_transformers import SentenceTransformer

embedding_model = SentenceTransformer("all-MiniLM-L6-v2")

embeddings = embedding_model.encode(chunks)

print(embeddings.shape)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:112: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.5k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

(81, 384)


81 chunks

384 numbers per chunk

Chunk 1 -> [0.12, -0.34, 0.78, ... 384 values]

Chunk 2 -> [0.45,  0.11, 0.91, ... 384 values]

Chunk 3 -> [...]

...

Chunk 81 -> [...]

In [10]:
!pip install -q chromadb

import chromadb

client = chromadb.Client()

collection = client.create_collection("rag_collection")

collection.add(
    embeddings=embeddings.tolist(),
    documents=chunks,
    ids=[f"chunk_{i}" for i in range(len(chunks))]
)

print("Stored", len(chunks), "chunks in ChromaDB")

Stored 81 chunks in ChromaDB


Why in RAG:

PDF → Chunks → Embeddings → ChromaDB → Retrieval

Embeddings are just vectors in memory right now.
ChromaDB stores them so we can later search for the most relevant chunks using a query.

Explanation:

- chromadb is a vector database.
- Client() creates a local ChromaDB instance.
- create_collection() creates a container for vectors and documents.
- embeddings.tolist() converts NumPy arrays into Python lists.
- documents=chunks stores the original text of each chunk.
- ids creates a unique identifier for every chunk.
- collection.add() saves everything into ChromaDB.

Output:

Stored 81 chunks in ChromaDB

Meaning:

Each chunk now contains:
1. Original text
2. 384-dimensional embedding
3. Unique ID

and is stored inside ChromaDB, ready for retrieval.

documents=chunks

means:

document 0 = chunk 0
document 1 = chunk 1
document 2 = chunk 2
...

ChromaDB stores the chunk text under the key "documents".

Syntax:

collection.add(...) calls the add() function.

Inside the brackets, we pass arguments using:

parameter_name=value

---
Example:

embeddings=...
documents=...
ids=...

---
Loop Part:

[f"chunk_{i}" for i in range(len(chunks))]

This is called a List Comprehension.

Step-by-step:

len(chunks)
→ 81

range(81)
→ 0,1,2,3,...,80

For each i:

f"chunk_{i}"

creates:

chunk_0
chunk_1
chunk_2
...
chunk_80

Final result:

[
  "chunk_0",
  "chunk_1",
  "chunk_2",
  ...
  "chunk_80"
]

---
Why IDs?

Each document stored in ChromaDB needs a unique identifier so it can be retrieved and managed later.

---
Stored Data:

ID        Document       Embedding
chunk_0   chunk text     vector
chunk_1   chunk text     vector
chunk_2   chunk text     vector
...





**Retrieval**

In [11]:
query = "What is Logistic Regression?"

query_embedding = embedding_model.encode(query).tolist()

results = collection.query(
    query_embeddings=[query_embedding],
    n_results=3
)

print(results["documents"][0])

['containing discrete or continuous variables. In this section we will primarily con-\nsider the case where Y is a boolean variable, in order to simplify notation. In the\nfinal subsection we extend our treatment to the case where Y takes on any finite\nnumber of discrete values.\nLogistic Regression assumes a parametric form for the distribution P(Y|X),\nthen directly estimates its parameters from the training data. The parametric\nmodel assumed by Logistic Regression in the case where Y is boolean is:', 'Logistic Regression is often referred to as adiscriminative classifier because\nwe can view the distribution P(Y|X) as directly discriminating the value of\nthe target value Y for any given instance X.\n• Logistic Regression is a linear classifier over X. The linear classifiers pro-\nduced by Logistic Regression and Gaussian Naive Bayes are identical in\nthe limit as the number of training examples approaches infinity, provided\nthe Naive Bayes assumptions hold. However, if these assumptio

Why in RAG:

Find the most relevant chunks for a user's question.

Explanation:
- Convert the query into an embedding.
- Search ChromaDB using that embedding.
- Return the top 3 similar chunks.

Output:
3 chunks most relevant to "What is Logistic Regression?"

**Context Creation.**

In [12]:
context = "\n\n".join(results["documents"][0])

print(context)

containing discrete or continuous variables. In this section we will primarily con-
sider the case where Y is a boolean variable, in order to simplify notation. In the
final subsection we extend our treatment to the case where Y takes on any finite
number of discrete values.
Logistic Regression assumes a parametric form for the distribution P(Y|X),
then directly estimates its parameters from the training data. The parametric
model assumed by Logistic Regression in the case where Y is boolean is:

Logistic Regression is often referred to as adiscriminative classifier because
we can view the distribution P(Y|X) as directly discriminating the value of
the target value Y for any given instance X.
• Logistic Regression is a linear classifier over X. The linear classifiers pro-
duced by Logistic Regression and Gaussian Naive Bayes are identical in
the limit as the number of training examples approaches infinity, provided
the Naive Bayes assumptions hold. However, if these assumptions do not

Copyr

Why in RAG:
LLMs need a single text context, not a list of chunks.

Explanation:
- Takes the retrieved chunks.
- Merges them into one text block.
- Stores it in the variable context.

Output:
A single text containing all retrieved chunks.

**Mini RAG**

In [16]:
question = "What is Logistic Regression?"

prompt = f"""
Answer the question using only the context below.

Context:
{context}

Question:
{question}
"""

print(prompt)


Answer the question using only the context below.

Context:
containing discrete or continuous variables. In this section we will primarily con-
sider the case where Y is a boolean variable, in order to simplify notation. In the
final subsection we extend our treatment to the case where Y takes on any finite
number of discrete values.
Logistic Regression assumes a parametric form for the distribution P(Y|X),
then directly estimates its parameters from the training data. The parametric
model assumed by Logistic Regression in the case where Y is boolean is:

Logistic Regression is often referred to as adiscriminative classifier because
we can view the distribution P(Y|X) as directly discriminating the value of
the target value Y for any given instance X.
• Logistic Regression is a linear classifier over X. The linear classifiers pro-
duced by Logistic Regression and Gaussian Naive Bayes are identical in
the limit as the number of training examples approaches infinity, provided
the Naive Bayes 

Why in RAG:
Combine retrieved context and user question into a prompt for the LLM.

Explanation:
- question stores the user's query.
- context contains the retrieved chunks.
- f-string inserts both into one prompt.
- This prompt will be sent to an LLM next.

Output:
A complete prompt containing context + question.

**send the prompt to an LLM.**

In [18]:
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

tokenizer = AutoTokenizer.from_pretrained("google/flan-t5-base")
model = AutoModelForSeq2SeqLM.from_pretrained("google/flan-t5-base")

inputs = tokenizer(prompt, return_tensors="pt")

outputs = model.generate(
    **inputs,
    max_new_tokens=100
)

answer = tokenizer.decode(outputs[0], skip_special_tokens=True)

print(answer)

Loading weights:   0%|          | 0/282 [00:00<?, ?it/s]

[transformers] The tied weights mapping and config for this model specifies to tie shared.weight to lm_head.weight, but both are present in the checkpoints with different values, so we will NOT tie them. You should update the config with `tie_word_embeddings=False` to silence this warning.


Logistic Regression is often referred to as adiscriminative classifier because we can view the distribution P(Y|X) as directly discriminating the value of the target value Y for any given instance X. • Logistic Regression is often referred to as adiscriminative classifier because we can view the distribution P(Y|X) as directly discriminating the value of the target value Y for any given instance X. •


Why in RAG:

Prompt → LLM → Answer

Explanation:

- Loads FLAN-T5 tokenizer.
- Loads FLAN-T5 model.
- Converts prompt into tokens.
- Generates an answer.
- Converts output tokens back to text.

Output:

A direct answer generated from the retrieved context.